### ISSUE TITLE
Cannot reproduce the test accuracy of GRAIN (ball-D)

### ISSUE THREAD

**Issue by DAIZHENWEI** [zwt233/Grain#1]
Hi, I tried to run your code in ./examples/Test.ipynb. But I cannot reproduce the test accuracy of GRAIN (ball-D). The performance is about ~3% lower than that is shown in your paper. Could you check whether the parameters are correct in your file? I attached the figure showing my results.
![Grain_1](https://user-images.githubusercontent.com/32122197/145511529-40c5c6e9-ec62-41b3-af24-f026bdf3c4b2.png)

**Comment by zwt233** [comment]
> Hi, I tried to run your code in ./examples/Test.ipynb. But I cannot reproduce the test accuracy of GRAIN (ball-D). The performance is about ~3% lower than that is shown in your paper. Could you check whether the parameters are correct in your file? I attached the figure showing my results. ![Grain_1](https://user-images.githubusercontent.com/32122197/145511529-40c5c6e9-ec62-41b3-af24-f026bdf3c4b2.png)

Thanks for your attention!
The test accuracy of the learning-based GNN model may be lower than validation accuracy, leading to unstable performance. Therefore, we report the mean test accuracy in our work. 
Note that the node selection process in Grain is parameter-free，and the unstable performance mainly comes from the training of the GCN model. We have added the early-stopping trick when training in the file early_stop_training.py. For fairness, all the baselines are trained by the same python file since selecting nodes and training the model are two independent stages.
We hope this will solve your problem.
![9ee464607a477269316d33c60a119e5](https://user-images.githubusercontent.com/12367386/145666202-5a8243b5-d4e9-4fd2-b250-8d7979e9128d.png)

**Comment by frankhlchi** [comment]
Hi Wt, thank you for making the code public. Based my understanding, the evaluation code in the early stopping.py:

> print('xxxxxxxxxx Evaluation begin xxxxxxxxxx')
> t_total = time.time()
> record = {}
> for i in range(500):
>     model = GCN(nfeat=features_GCN.shape[1],
>             nhid=hidden_size,
>             nclass=labels.max().item() + 1,
>             dropout=0.85)
>     model.cuda()
>     early_stopping = EarlyStopping(patience = 10)
>     optimizer = optim.Adam(model.parameters(),
>                            lr=0.05, weight_decay=5e-4)
>     for epoch in range(400):
>         train(epoch,model,record)
>         if early_stopping.early_stop==True:
>             break
> 
> bit_list = sorted(record.keys())
> bit_list.reverse()
> for key in bit_list[:10]:
>     value = record[key]
>     print(round(key,3),round(value,3))
> print('xxxxxxxxxx Evaluation end xxxxxxxxxx')
> 

This seems to show the best top 10 performers among 500 runs instead of average. Is this correct? Thank you.

**Comment by pkuwyx** [comment]
> Hi Wt, thank you for making the code public. Based my understanding, the evaluation code in the early stopping.py:
> 
> > print('xxxxxxxxxx Evaluation begin xxxxxxxxxx')
> > t_total = time.time()
> > record = {}
> > for i in range(500):
> > model = GCN(nfeat=features_GCN.shape[1],
> > nhid=hidden_size,
> > nclass=labels.max().item() + 1,
> > dropout=0.85)
> > model.cuda()
> > early_stopping = EarlyStopping(patience = 10)
> > optimizer = optim.Adam(model.parameters(),
> > lr=0.05, weight_decay=5e-4)
> > for epoch in range(400):
> > train(epoch,model,record)
> > if early_stopping.early_stop==True:
> > break
> > bit_list = sorted(record.keys())
> > bit_list.reverse()
> > for key in bit_list[:10]:
> > value = record[key]
> > print(round(key,3),round(value,3))
> > print('xxxxxxxxxx Evaluation end xxxxxxxxxx')
> 
> This seems to show the best top 10 performers among 500 runs instead of average. Is this correct? Thank you.

Thank you for your attention to our work. We do not simply select the best result in validation set. In fact, we select the top 10 validation acc to ensure  the stability of early-stopping(we use early-stopping to make sure that the result of validation set and test set can be nearly same, or later will be less than former with a significant gap).
Do not worry about the justice, all the baseline use the same technique(like AGE,ANRMAB). That is why both AGE and ANRMAB seem better than original paper report.

**Comment by frankhlchi** [comment]
Thank you for your swift response. I was wondering if it is possible to share or upload the generated "GRAIN(ball-D)_r0.05_cora_selected_nodes.json" file to the repository? I got the file by running the Ball.py; just want to make sure I got the correct results. Thanks!

**Comment by Gaop970222** [comment]
> > Hi, I tried to run your code in ./examples/Test.ipynb. But I cannot reproduce the test accuracy of GRAIN (ball-D). The performance is about ~3% lower than that is shown in your paper. Could you check whether the parameters are correct in your file? I attached the figure showing my results. ![Grain_1](https://user-images.githubusercontent.com/32122197/145511529-40c5c6e9-ec62-41b3-af24-f026bdf3c4b2.png)
> 
> Thanks for your attention! The test accuracy of the learning-based GNN model may be lower than validation accuracy, leading to unstable performance. Therefore, we report the mean test accuracy in our work. Note that the node selection process in Grain is parameter-free，and the unstable performance mainly comes from the training of the GCN model. We have added the early-stopping trick when training in the file early_stop_training.py. For fairness, all the baselines are trained by the same python file since selecting nodes and training the model are two independent stages. We hope this will solve your problem. ![9ee464607a477269316d33c60a119e5](https://user-images.githubusercontent.com/12367386/145666202-5a8243b5-d4e9-4fd2-b250-8d7979e9128d.png)

Hi, thanks for your work. When I reproduce your code with early stop training, I notice that I still cannot get the similar reults. The below is the code and correspoding results:
```python
import numpy as np
import pickle as pkl
import networkx as nx
import scipy.sparse as sp
import torch
import random
import copy
import sys
import os
import time
import argparse
import json
import numpy as np
import numpy.linalg as la
import torch.nn.functional as F
import torch.optim as optim
import torch.nn as nn
import pandas as pd
from scipy.sparse import csgraph
from torch.backends import cudnn
from torch.optim import lr_scheduler
from utils import *
from graphConvolution import *

num_node = 2708
num_coreset = 140
hidden_size = 128
dmax = np.ones(num_node)
gamma = 1
radium = 0.05

class GCN(nn.Module):
    def __init__(self, nfeat, nhid, nclass, dropout):
        super(GCN, self).__init__()

        self.gc1 = GraphConvolution(nfeat, nhid,bias=True)
        self.gc2 = GraphConvolution(nhid, nclass,bias=True)
        self.dropout = dropout

    def forward(self, x, adj):
        x = F.dropout(x, self.dropout, training=self.training)
        x = F.relu(self.gc1(x, adj))
        x = F.dropout(x, self.dropout, training=self.training)
        x = self.gc2(x, adj)
        return x

def train(epoch, model,record):

    model.train()
    optimizer.zero_grad()
    output = model(features_GCN, adj)
    loss_train = F.cross_entropy(output[idx_train], labels[idx_train])
    acc_train = accuracy(output[idx_train], labels[idx_train])
    loss_train.backward()
    optimizer.step()
    model.eval()
    output = model(features_GCN, adj)

    loss_val = F.nll_loss(output[idx_val], labels[idx_val])
    acc_val = accuracy(output[idx_val], labels[idx_val])
    loss_test = F.nll_loss(output[idx_test], labels[idx_test])
    acc_test = accuracy(output[idx_test], labels[idx_test])
    record[acc_val.item()] = acc_test.item()


def get_receptive_fields_dense(cur_neighbors, selected_node, weighted_score):
    receptive_vector=((cur_neighbors+adj_matrix2[selected_node])!=0)+0
    count=weighted_score.dot(receptive_vector)
    return count

def get_current_neighbors_dense(cur_nodes):
    if np.array(cur_nodes).shape[0]==0:
        return 0
    neighbors=(adj_matrix2[list(cur_nodes)].sum(axis=0)!=0)+0
    return neighbors

def get_max_nnd_node_dense(idx_used,high_score_nodes,min_distance):
    max_receptive_node = 0
    max_total_score = 0
    cur_neighbors=get_current_neighbors_dense(idx_used)
    for node in high_score_nodes:
        receptive_field=get_receptive_fields_dense(cur_neighbors,node,num_ones)
        node_distance = distance_aax[node,:]
        node_distance = np.where(node_distance<min_distance,node_distance,min_distance)
        node_distance = dmax - node_distance
        distance_score = node_distance.dot(num_ones)
        total_score = receptive_field/num_node+gamma*distance_score/num_node
        if total_score > max_total_score:
            max_total_score = total_score
            max_receptive_node = node
    return max_receptive_node

def aug_normalized_adjacency(adj):
    adj = adj + sp.eye(adj.shape[0])
    adj = sp.coo_matrix(adj)
    row_sum = np.array(adj.sum(1))
    d_inv_sqrt = np.power(row_sum, -0.5).flatten()
    d_inv_sqrt[np.isinf(d_inv_sqrt)] = 0.
    d_mat_inv_sqrt = sp.diags(d_inv_sqrt)
    return d_mat_inv_sqrt.dot(adj).dot(d_mat_inv_sqrt).tocoo()


def compute_distance(_i,_j):
    return la.norm(features_aax[_i,:]-features_aax[_j,:])

class EarlyStopping:
    def __init__(self, patience=7, delta=0):
        self.patience = patience
        self.counter = 0
        self.best_score = None
        self.early_stop = False
        self.val_loss_min = np.Inf
        self.delta = delta
    def __call__(self, val_loss):

        score = -val_loss

        if self.best_score is None:
            self.best_score = score
        elif score < self.best_score + self.delta:
            self.counter += 1
            if self.counter >= self.patience:
                self.early_stop = True
        else:
            self.best_score = score
            self.counter = 0

adj, features, labels, idx_train, idx_val, idx_test = load_data(dataset="cora")
features_GCN = copy.deepcopy(features)
features_GCN = torch.FloatTensor(features_GCN).cuda()

num_zeros = np.zeros(num_node)
num_ones = np.ones(num_node)
idx_val = list(idx_val.cpu())
idx_test = list(idx_test.cpu())
idx_avaliable = list()
for i in range(num_node):
    if i not in idx_val and i not in idx_test:
        idx_avaliable.append(i)

adj = aug_normalized_adjacency(adj)
adj_matrix = torch.FloatTensor(adj.todense()).cuda()
adj_matrix2 = torch.mm(adj_matrix,adj_matrix).cuda()
features = features.cuda()
features_aax = np.array(torch.mm(adj_matrix2,features).cpu())
adj_matrix2 = np.array(adj_matrix2.cpu())

distance_aax = np.zeros((num_node,num_node))
for i in range(num_node-1):
    for j in range(i+1,num_node):
        distance_aax[i][j] = compute_distance(i,j)
        distance_aax[j][i] = distance_aax[i][j]
dis_range = np.max(distance_aax) - np.min(distance_aax)
distance_aax = (distance_aax - np.min(distance_aax))/dis_range

print('GRAIN(NN-D) node selection start')
min_distance = np.ones(num_node)
idx_train_nnd = []
idx_avaliable_temp = copy.deepcopy(idx_avaliable)
count = 0
while True:
    max_reception_node = get_max_nnd_node_dense(idx_train_nnd,idx_avaliable_temp,min_distance)
    idx_train_nnd.append(max_reception_node)
    idx_avaliable.remove(max_reception_node)
    idx_avaliable_temp.remove(max_reception_node)
    count += 1
    max_node_distance = distance_aax[max_reception_node,:]
    min_distance = np.where(min_distance<max_node_distance,min_distance,max_node_distance)
    if count >= num_coreset:
        break

print('GRAIN(NN-D) node selection finished')

print('GRAIN(ball-D) node selection start')
balls = np.zeros((num_node,num_node))
balls_dict=dict()
covered_balls = set()
for i in range(num_node):
    for j in range(num_node):
        if distance_aax[i][j] <= radium:
            balls[i][j]=1

idx_avaliable_tmp = copy.deepcopy(idx_avaliable)
for node in idx_avaliable_tmp:
    neighbors_tmp = get_current_neighbors_dense([node])
    neighbors_tmp = neighbors_tmp[:,np.newaxis]
    dot_result = np.matmul(balls,neighbors_tmp).T
    tmp_set = set()
    for i in range(num_node):
        if dot_result[0,i]!=0:
            tmp_set.add(i)
    balls_dict[node]=tmp_set

count = 0
idx_train_balld = []
while True:
    ball_num_max = 0
    node_max = 0
    for node in idx_avaliable_tmp:
        tmp_num = len(covered_balls.union(balls_dict[node]))
        if tmp_num > ball_num_max:
            ball_num_max = tmp_num
            node_max = node
    res_ball_num = num_node - ball_num_max
    count+=1
    idx_train_balld.append(node_max)
    idx_avaliable_tmp.remove(node_max)
    covered_balls = covered_balls.union(balls_dict[node_max])
    if count >= num_coreset or res_ball_num==0:
        break

print('GRAIN(ball-D) node selection finished')


adj = sparse_mx_to_torch_sparse_tensor(adj).float().cuda()
labels = labels.cuda()
idx_val = torch.LongTensor(idx_val).cuda()
idx_test = torch.LongTensor(idx_test).cuda()

print('xxxxxxxxxx GRAIN(NN-D) Evaluation begin  xxxxxxxxxx')
idx_train = torch.LongTensor(idx_train_nnd).cuda()
record = {}
early_stopping = EarlyStopping(patience = 10)
model = GCN(nfeat=features_GCN.shape[1],
        nhid=hidden_size,
        nclass=labels.max().item() + 1,
        dropout=0.85)
model.cuda()
optimizer = optim.Adam(model.parameters(),
                       lr=0.05, weight_decay=5e-4)
for epoch in range(400):
    train(epoch,model,record)
    if early_stopping.early_stop==True:
        break

bit_list = sorted(record.keys())
bit_list.reverse()
for key in bit_list[:10]:
    value = record[key]
    print(key,value)
print('xxxxxxxxxx GRAIN(NN-D) Evaluation end  xxxxxxxxxx')


print('xxxxxxxxxx GRAIN(ball-D) Evaluation begin  xxxxxxxxxx')
idx_train = torch.LongTensor(idx_train_balld).cuda()
record = {}
model = GCN(nfeat=features_GCN.shape[1],
        nhid=hidden_size,
        nclass=labels.max().item() + 1,
        dropout=0.85)
model.cuda()
optimizer = optim.Adam(model.parameters(),
                       lr=0.05, weight_decay=5e-4)
early_stopping = EarlyStopping(patience = 10)
for epoch in range(400):
    train(epoch,model,record)
    if early_stopping.early_stop==True:
        break

bit_list = sorted(record.keys())
bit_list.reverse()
for key in bit_list[:10]:
    value = record[key]
    print(key,value)
print('xxxxxxxxxx GRAIN(ball-D) Evaluation end  xxxxxxxxxx')
```


<img width="447" height="741" alt="Image" src="https://github.com/user-attachments/assets/b37cc886-f9b9-40a8-afec-1265a8a12d27" />

If you are available, could you upload the code with early stopping trick? Thank you very much!!! 

Best Regards
